### ISSUE TITLE
NavigationAgent2D doesn't find shortest past with NavigationLink2D of cost zero

### ISSUE THREAD

**Issue by sandygk** [godotengine/godot#84388]
### Godot version

Godot v4.1.2.stable 

### System information

Godot v4.1.2.stable - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Laptop GPU (NVIDIA; 31.0.15.3699) - AMD Ryzen 9 6900HS with Radeon Graphics (16 Threads)

### Issue description

I have a simple scene with a NaviationAgent2D and a NavigationLink2D with travel cost 0 to simulate a teleportation portal, and a Target. The path the agent finds is a straight line to the target, which is suboptimal, since going through the link/portal is shorter because it has a travel cost of 0. 

![image](https://github.com/godotengine/godot/assets/6596836/2cedb134-99e9-47dc-9882-a96f3f631cfa)

### Steps to reproduce

Run the minimal reproduction project and notice the path that it finds is not the correct one:

### Minimal reproduction project

[nav-agent-2d-bug.zip](https://github.com/godotengine/godot/files/13245458/nav-agent-2d-bug.zip)

[nav-agent-2d-bug-different-polygons.zip](https://github.com/godotengine/godot/files/13251862/nav-agent-2d-bug-different-polygons.zip)

**Comment by smix8** [comment]
> The path the agent finds is a straight line to the target, which is suboptimal

I get why you see it that way but from the perspective of the involved pathfinding this result is perfection. It found that both positions are on the same polygon so moved in a straight line to the target position.

That setup will not work with pure A* pathfinding. You would need to query a link map first. It would not even work with a different custom pathfinding algorithm, e.g. Dijkstra, because both start and target polygon are the same, so the result is always a straight path line without a link map that gets queried first.

A* pathfinding uses a greedy heuristic with euclidean distance towards the target position, the heuristic part is core why A* is so efficient but it works against it to find alternative and objectively "better" path options that do not follow the direction to the target position. Since the path to the target is unobstructed and on the same polygon it finds a direct path corridor and returns that path. At that point the job is done as far as the A* pathfinding is concerned and it does an early exit, it does not search any other polygons for "better paths". So from the perspective of A* pathfinding this is a perfect optimal path.

Links with A* will only be used reliably when they are the only connection between navigation mesh islands or the polygon surface cost of walking with polygons in the direction of the target is far height than the link. A link that is left or right of the direction will only be used by chance with enough obstruction in the direction and a link that is the opposite direction of the target will only be found when all other polygon search options are exhausted. That is the nature of greedy A* pathfinding.

**Comment by sandygk** [comment]
I am not sure how the mesh is transformed into a graph to pass to the pathfinding algorithm, the simplest way I can think of is that we pick points from the mesh to create the nodes (let's say at 1px distance for instance) and if 2 points are next to each other, then we create an edge to connect them.  If this is the case, then the above statement is incorrect:

>That setup will not work with pure A* pathfinding. You would need to query a link map first. It would not even work with a different custom pathfinding algorithm, e.g. Dijkstra, because both start and target polygon are the same, so the result is always a straight path line without a link map that gets queried first.

In the case of Dijkstra, is guaranteed to return the optimal path as long as the weights (or costs) are not negative, which the engine enforces with `travel cost >= 0`.

In the case of A* it is also guaranteed to return the shortest path as long as the [heuristic is admissible](https://en.wikipedia.org/wiki/Admissible_heuristic), which I don't see why we would pick one that isn't.

Thanks for the help, I'll definitely take a look at your suggestion, but I am pretty sure this is a bug in the pathfinding algorithm that we should fix in the Engine

**Comment by sandygk** [comment]
I looked at the code and I see what you mean, the code assumes that if bot the start and end points are in the same convex polygon then the solution is the straight line, but this assumption is incorrect as soon as you add links, so the code need to be updated, otherwise it doesn't return the shortest path as illustrated in the example above.

But there is also another issue going on, that has nothing to do with being on the same polygon, see the following image:

<img width="962" alt="image" src="https://github.com/godotengine/godot/assets/6596836/7471ddd6-4879-4ab7-81c6-09048a606b9b">

The agent still ignores the link, even though the path it takes is way longer, in this case, the start and end point are not in the same polygon

**Comment by sandygk** [comment]
@smix8, I just looked at the code and I understand what you mean by:

> That setup will not work with pure A* pathfinding.

Even though theoretically A* returns the shortest path when the heuristic function is [admissible](https://en.wikipedia.org/wiki/Admissible_heuristic#:~:text=An%20admissible%20heuristic%20is%20used,of%20reaching%20the%20goal%20state.), we are using the Euclidean distance as a heuristic in our implementation, and this doesn't really work with links since links pretty much make the Geometry not Euclidean, and the heuristic is no longer admissible.

But I am still pretty sure it should work with Dijkstra, I don't see why it wouldn't.

I would investigate to see if we can use any other heuristic that is still admissible for a geometry with links. I'll also try making the `h` function always return `0` which pretty much turns A* into Dijkstra and check if it works with that change. Just to make sure.

**Comment by sandygk** [comment]
Yup, I modify the A* to turn it into Dijkstra and that basically solve the issue, obviously it takes longer to compute, but for a small case like this is not really significant:

![image](https://github.com/godotengine/godot/assets/6596836/96dfc9fe-8390-4d3b-997f-71e6626c8fe7)

The modification to turn A* into Dijkstra are very simple, for the most part is just removing this line:

https://github.com/godotengine/godot/blob/5ee983188de97ae027f9b9c1443438063f708a7e/modules/navigation/nav_map.cpp#L364

That would basically set the heuristic function to 0

We also need to fix the corner case where the start and end are in the same polygon, but that seems relatively simple. 

What I would suggest is to add Dijkstra as an alternative Pathfinding Algorithm since A* doesn't really work well with links, but Dijkstra does.

We even have the UI pretty much ready for that:

![image](https://github.com/godotengine/godot/assets/6596836/4281ea1b-1778-4e89-8c70-f49b0b43bb75)

**Comment by smix8** [comment]
Dijkstra will fix your specific test project because there are only ~5 polygons in the direction of the link and 10+ in the direction to the target position. It will not fix the issue in general.

It would still be a mostly circular exploration pattern of polygons, so when link and target position are placed at mostly equal polygon numbers the link gets overlooked because the pathfinding will do an early exist as soon as the target position polygon is found.

So what also needs to change is on top of the Dijkstra pathfinding option the early exit also needs to be made optional / with a polygon search cap. E.g. -1 is no cap, the search will go until it runs out of polygons, and any >0 positive number will be the number of additional polygons searched after a path search finds the polygon with the target position.

If you want to work and test around with this be my guest. In general it is not that many of those issues are not known but more that they have been rescheduled to a later point to be paired with planned reworks and performance improvements as battles have to be picked carefully around a navigation team the size of 0.5.

Since Godot 4 is release one constrain is that changes that worsen the pathfinding for everyone to fix specific niche issues are only acceptable with general performance gains for a net positive. Especially the begin_poly == end_poly branch is very hot, as user projects love to spam path searches while staying on the same polygon forever.

**Comment by sandygk** [comment]
> It would still be a mostly circular exploration pattern of polygons, so when link and target position are placed at mostly equal polygon numbers the link gets overlooked because the pathfinding will do an early exist as soon as the target position polygon is found.
>
>So what also needs to change is on top of the Dijkstra pathfinding option the early exit also needs to be made optional / with a polygon search cap. E.g. -1 is no cap, the search will go until it runs out of polygons, and any >0 positive number will be the number of additional polygons searched after a path search finds the polygon with the target position.

I see, that makes sense. I can do that.

> If you want to work and test around with this be my guest. In general it is not that many of those issues are not known but more that they have been rescheduled to a later point to be paired with planned reworks and performance improvements as battles have to be picked carefully around a navigation team the size of 0.5.

I'll give it a try, I'll post here if I make substantial progress.

> Since Godot 4 is release one constrain is that changes that worsen the pathfinding for everyone to fix specific niche issues are only acceptable with general performance gains for a net positive. Especially the begin_poly == end_poly branch is very hot, as user projects love to spam path searches while staying on the same polygon forever.

That makes sense, but considering this will be a different pathfinding algorithm and the user has to opt-in it won't really affect any user that doesn't select it

**Comment by sandygk** [comment]
>Dijkstra will fix your specific test project because there are only ~5 polygons in the direction of the link and 10+ in the direction to the target position. It will not fix the issue in general.

I think Dijkstra still works, it's just that we need to be clever about it. Here is an idea, what if we create polygons for the start and end points. Basically the polygons will have an area of 0, and they will be the same point repeated 3 times, similar to the hack we use for turning links into polygons. I am pretty sure this will take care of all the issues we have mentioned and seems pretty straightforward to implement.

**Comment by sandygk** [comment]
Another approach could be to modify the A* heuristic to allow for links. 

Specifically, the issue arises with reachable Non-Euclidean links that enable faster travel than what's possible in Euclidean geometry, making the Euclidean distance heuristic inadmissible. During link polygon creation, we can identify and mark Non-Euclidean links. When listing reachable polygons, we then check for reachable Non-Euclidean links and **only then apply the modifications below**.

1. Utilize the start and end dummy polygons as previously mentioned.
  
2. Adjust the heuristic to account for Non-Euclidean links, reformulated as:

$\ h(n) = \min(d(n,\text{goal}), \min_{l \in \text{Non-Euclidean links}} (d(n,l) + \text{entry cost}(l) + d(l,\text{goal})) ) \$


Here, $d(n,\text{{goal}})$ and $d(n,l)$ are computed using Euclidean distance. However, $d(l,\text{{goal}})$ must consider potential traversal through other Non-Euclidean links, requiring precomputation.

3. Precompute shortest distances between each Non-Euclidean link and the goal using Dijkstra's algorithm:

    - Construct a graph where nodes represent Non-Euclidean link endpoints and the goal. 
    - Connect all link exits to all link entries, with the cost being the sum of the Euclidean distance plus the entry cost.
    - Also connect all link exits to the goal, with the cost being the Euclidean distance from the exit to the goal.
    - Connect all entries to their respective exits, with the cost being the product of the link length and its distance.
    - Execute Dijkstra's algorithm from the goal to each entry, precomputing $d(n,\text{{goal}})$ for each reachable Non-Euclidean link.


Thoughts?

**Comment by smix8** [comment]
The hack of the navigation links with their mini polygon is not a good example to follow, it only exists because at that time things were under Godot 4.0 feature freeze and release pressure. There was some informal agreement and merge condition to fix it after release for 4.1 but then the link contributor disappeared. The navigation edge connection calculation did that part better in that they did not add additional polygons  (and other performance things worse).

Hacking with a corrupted zero surface polygon is not the way to go. The navigation map would throw such things rightfully back at users. The navigation mesh system is build around convex polygons with an expected surface area that can also be queried by users.

There was some discussion about adding waypoints to the NavigationServer to migrate the AStar classes to the server. This would also add a "point" alternative to polygons which could also be used for link positions but no work has started on this.

Pre-computing could be an additional option for static scenes but there also need to be a general solution that works for runtime changes. You need to consider that users will commonly do stuff like this in their release projects:
![linkoveruse](https://github.com/godotengine/godot/assets/52464204/90a865ba-0a32-4025-8c64-81c76c2dd5e1)
This requires a hierarchical approach so chunks of the map can be computed in the background which is currently not an option with how the navigation map syncs its map state.

**Comment by sandygk** [comment]
I just realized that my way of thinking about Euclidean vs Non-Euclidean is not the right framework. Euclidean distance assumes constant travel cost, the problem here is not with links being Non-Euclidean, the problem appears when we have different travel costs involved. 

Here is an example:

![image](https://github.com/godotengine/godot/assets/6596836/ffe198c5-6355-4212-b78b-27ca32951fdf)
Let's say the red area is mud and really slow to walk (high travel cost) and the green area is grass and fast to walk (low travel cost), and the purple lines are links to connect both areas.

Our current heuristic won't find the proper path which is walking through the grass, but not because the link is not Euclidean, the link could have the same walking speed as the mud, and it will still be better to use the grass. So the suggestion I made in [this comment](https://github.com/godotengine/godot/issues/84388#issuecomment-1793416859) won't really fix the issue for the general case.

**Comment by sandygk** [comment]
So to summarize, there are two issues:

1. **The heuristic used is non-admissible**: This is a requirement to find the shortest path. A quick solution for that is, in this line:
[cost += (np->entry.distance_to(end_point) * np->poly->owner->get_travel_cost());](https://github.com/godotengine/godot/blob/5ee983188de97ae027f9b9c1443438063f708a7e/modules/navigation/nav_map.cpp#L364)
instead of using the travel cost of the current polygon, is to use the lowest travel of the reachable regions. 
For an extra bit of performance we can also add the cost of entering the goal's region if it's a different region than the current.

2. **The stopping condition is incorrect**: it assumes that if we reach the goal's polygon we can go directly to the goals, but this also yields suboptimal paths. For this the solution is to wrap the goal into its own zero surface polygon. That we if this is polygon reached, we can guarantee that the shortest path was found.

I'll see if I can put together a PR with the proposed solutions, so we can discuss in the PR in more concrete terms

**Comment by sandygk** [comment]
I am working on the PR but I found another issue with the following logic:

https://github.com/godotengine/godot/blob/5ee983188de97ae027f9b9c1443438063f708a7e/modules/navigation/nav_map.cpp#L249-L275

This logic basically prevents us from entering the same polygon twice, but sometimes the shortest path has to do exactly that as shown in the example mentioned [here](https://github.com/godotengine/godot/issues/84388#issuecomment-1795315626). I am trying to think how to work around that

**Comment by sandygk** [comment]
I'm working on the PR, but I am a bit stuck at the moment. The current implementation assumes we should not visit the same polygon multiple times, but this assumption is incorrect if we want to find the shortest path. Sometimes the shortest path may re-enter the same polygon. I am trying to find a way to allow for this without having a huge impact in the performance.

**Comment by sandygk** [comment]
@smix8, so have the changes working on a POC branch in my local, and they fix the above-mentioned issues, during the implementation I found a few issues with the current implementation and what I would suggest doing is a small refactoring to address the following issues, but I wanted to run it by you first, if you have a chance:

So what I am proposing is:

1.  **Create the graph using the centroid of the polygons as nodes, instead of the using the polygons themselves**: This seems to be the most popular approach from the research I made, it seems like Unity does this according to [their documentation](https://docs.unity3d.com/Packages/com.unity.ai.navigation@1.1/manual/AreasAndCosts.html). It would look like this before the string pulling optimization:
![image](https://github.com/godotengine/godot/assets/6596836/503c84fb-52ab-467c-bc63-f286615149b7)
This approach simplifies the code since we will be dealing with actual nodes, that way we can add links as two nodes and represent the start and end points as nodes as well. Right now since the algorithm deals with polygons instead, we need to turn the links into polygons, and the same has to be done for the end point to address one of the issues mentioned on a previous comment, this solution is hacky and error-prone. 

2. **Refactor the way we represent connection across nodes**: Currently the connections are stored inside the edges of the polygon, this is a bad model to represent the connections in the mesh as soon as you introduce links, a link endpoint usually is located inside a polygon, not on the edge, so storing the connection between the link endpoint and the polygon in one of the edges of the polygon doesn't really make much sense. Connections can still have a reference to its owner, whether it is a polygon, a link, or a point. It should also have the pathway start and end, which is used in the path optimization.

3. **Refactor `NavigationPoly` to `VisitedNode`**: This class just represents a node that has being visited by the algorithm. It may still store pretty much the same information, just that the `poly` field should really be `owner` and store either a polygon a link or a point

4. **Fix the heuristic function to use the minimum reachable cost instead of the cost of the current polygon**: This will fix the vast majority of the problems introduced when there are multiple travel costs. The minimum reachable cost could be precomputed on the `sync` function. This introduces a performance issue when the minimum reachable cost is 0 since it will basically turn _A*_ into _Dijkstra_, for this I suggest we give a warning to the user that setting travel cost to 0 may affect performance or don't allow them to do it.

5. **Add the start and end point to the graph to fix current stopping condition**: Currently the search stops when it reaches the last polygon, this can return suboptimal results. This is particularly bad when both start and end are in the same polygon, adding the end point fixes that.

6. **Add enter cost to the nodes before entering the queue**: We currently add it after we get it out of the queue which affect the result, this is a small fix. Note, I am saying queue here in the theoretical sense, it really is a list.

7.  **Maybe split the algorithm into multiple functions for readability** : This is just a suggestion, but currently we have a really long function to do the entire process, I think it could be split into several functions to improve readability. Stuff like: finding start and end polygon, the search, the path optimization, etc. can be moved to its own function without affecting performance.

**Comment by QbieShay** [comment]
Hey :wave: 

First of all, thank you a lot for taking the time to look into this and even coding your own solution. It's awesome to see so much interest in improving the engine!

Regarding what you propose, I think there's a mismatch between the way we work and what you are proposing. As of now we are at 4.x stable, so compatibility breakage is out of the question unfortunately. By compatibility breakage, I don't only mean changing API, but also significantly change behaviours that users rely on.

Lastly, big refactors are hard to review and tackle, and it would be easier to approach it with smaller changes. While I understand that this is not ideal in the prospect of clean well designed code, it is the reality of maintaining a software that thousands of people use daily. And I know how painful it is to work this way, I've just done a massive refactor of particles and I had to make loads of compromises to keep things working with the previous implementation. So, I understand that what I'm saying here is not pleasant.

All of this to say that the best way to improve navigation would be to find ways to offer what you're proposing without changing existing behaviours and by making smaller changes. I hope this isn't discouraging for you, I want to reiterate that your energy to look into this is really really appreciated :pray: 

Feel free also to join our developer chat, if you're not there already https://chat.godotengine.org/
I find it easier sometimes to exchange some messages rather than having very async conversation on github.

**Comment by sandygk** [comment]
Hi @QbieShay, 

First, thank you for the encouragement and the kind words.

What you are saying makes a lot of sense. I am working on a PR to addresses some of the issues I mentioned above. The main goal would be to remove the link polygon hack, where we create artificial polygons for the nav links. But it also addresses a couple of other issues and improves the code readability. I'll make a detailed list of all the changes once the PR is ready. I was working on it for a while no, so is a bit extensive, but I'll just stop adding changes and go with want I have so far in order to keep it of a reasonable size. I'll add a list of what's next for some changes I think we should tackle once the PR is merged. 

I am currently a bit stuck on the logic to optimize the path, I'm reading about that algorithm to be able to address some changes I need to make there. Once/If I manage to resolve this issue, I'll open the PR.

I'll keep you posted, thanks for the help

**Comment by sandygk** [comment]
Hey @QbieShay , @smix8 , to iterate over the process of addressing the above-mentioned issues, I created the following PR to [Enhancing Readability and Reducing Complexity in NavMap sync and get_path Methods](https://github.com/godotengine/godot/pull/85038). The PR only aims to reduce some of the code complexity and improve its readability while preserving its functionality. I found that to be a good first step. So I submit it here for your consideration, if you find the time. Thanks for all the help!

**Comment by sandygk** [comment]
I created the following issues to track some of the issues mentioned here individually, so we can close this one
https://github.com/godotengine/godot-proposals/issues/8496
https://github.com/godotengine/godot-proposals/issues/8497

**Comment by sandygk** [comment]
Closing in favor of the issues linked on the previous comment

**Comment by sandygk** [comment]
I created the following issues to track some of the issues mentioned here individually, so we can close this one

https://github.com/godotengine/godot-proposals/issues/8496
https://github.com/godotengine/godot-proposals/issues/8497
https://github.com/godotengine/godot/issues/85238
https://github.com/godotengine/godot-proposals/issues/8499
https://github.com/godotengine/godot-proposals/issues/8501
